File: backend/app/models/master_support.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.master_database import MasterBase


class MasterSupportTicket(MasterBase):
    __tablename__ = "master_support_tickets"

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(
        ForeignKey("companies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    company_code: Mapped[str] = mapped_column(String(80), nullable=False, index=True)
    company_name: Mapped[str] = mapped_column(String(180), nullable=False)
    module: Mapped[str] = mapped_column(String(40), nullable=False, default="outro")
    priority: Mapped[str] = mapped_column(String(20), nullable=False, default="normal")
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="aberto")
    subject: Mapped[str] = mapped_column(String(180), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    requester_user_id: Mapped[int | None] = mapped_column(Integer)
    requester_name: Mapped[str | None] = mapped_column(String(150))
    requester_email: Mapped[str | None] = mapped_column(String(180))
    customer_attachments_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
    )
    assigned_master_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("master_users.id", ondelete="SET NULL"),
        index=True,
    )
    first_response_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    last_message_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    company = relationship("Company")
    assigned_master_user = relationship("MasterUser")
    messages = relationship(
        "MasterSupportMessage",
        back_populates="ticket",
        cascade="all, delete-orphan",
        order_by="MasterSupportMessage.created_at",
    )

    @property
    def assigned_master_user_name(self) -> str | None:
        return self.assigned_master_user.name if self.assigned_master_user else None

    @property
    def assigned_master_user_email(self) -> str | None:
        return self.assigned_master_user.email if self.assigned_master_user else None


class MasterSupportMessage(MasterBase):
    __tablename__ = "master_support_messages"

    id: Mapped[int] = mapped_column(primary_key=True)
    ticket_id: Mapped[int] = mapped_column(
        ForeignKey("master_support_tickets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    author_type: Mapped[str] = mapped_column(String(20), nullable=False)
    author_user_id: Mapped[int | None] = mapped_column(Integer)
    author_name: Mapped[str | None] = mapped_column(String(150))
    author_email: Mapped[str | None] = mapped_column(String(180))
    body: Mapped[str] = mapped_column(Text, nullable=False)
    attachment_url: Mapped[str | None] = mapped_column(Text)
    attachment_name: Mapped[str | None] = mapped_column(String(220))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    ticket = relationship("MasterSupportTicket", back_populates="messages")

File: admin_app/admin_flutter/lib/models/master_support.dart
class MasterSupportMessage {
  const MasterSupportMessage({
    required this.id,
    required this.ticketId,
    required this.authorType,
    required this.body,
    required this.createdAt,
    this.authorName,
    this.authorEmail,
    this.attachmentUrl,
    this.attachmentName,
  });

  final int id;
  final int ticketId;
  final String authorType;
  final String? authorName;
  final String? authorEmail;
  final String body;
  final String? attachmentUrl;
  final String? attachmentName;
  final DateTime createdAt;

  factory MasterSupportMessage.fromJson(Map<String, dynamic> json) {
    return MasterSupportMessage(
      id: json['id'] as int,
      ticketId: json['ticket_id'] as int,
      authorType: json['author_type'].toString(),
      authorName: json['author_name']?.toString(),
      authorEmail: json['author_email']?.toString(),
      body: json['body'].toString(),
      attachmentUrl: json['attachment_url']?.toString(),
      attachmentName: json['attachment_name']?.toString(),
      createdAt: DateTime.parse(json['created_at'].toString()),
    );
  }
}

class MasterSupportTicket {
  const MasterSupportTicket({
    required this.id,
    required this.companyId,
    required this.companyCode,
    required this.companyName,
    required this.module,
    required this.priority,
    required this.status,
    required this.subject,
    required this.description,
    required this.lastMessageAt,
    required this.createdAt,
    required this.updatedAt,
    required this.messages,
    this.customerAttachmentsEnabled = false,
    this.requesterName,
    this.requesterEmail,
    this.assignedMasterUserName,
    this.assignedMasterUserEmail,
  });

  final int id;
  final int companyId;
  final String companyCode;
  final String companyName;
  final String module;
  final String priority;
  final String status;
  final String subject;
  final String description;
  final bool customerAttachmentsEnabled;
  final String? requesterName;
  final String? requesterEmail;
  final String? assignedMasterUserName;
  final String? assignedMasterUserEmail;
  final DateTime lastMessageAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<MasterSupportMessage> messages;

  factory MasterSupportTicket.fromJson(Map<String, dynamic> json) {
    return MasterSupportTicket(
      id: json['id'] as int,
      companyId: json['company_id'] as int,
      companyCode: json['company_code'].toString(),
      companyName: json['company_name'].toString(),
      module: json['module'].toString(),
      priority: json['priority'].toString(),
      status: json['status'].toString(),
      subject: json['subject'].toString(),
      description: json['description'].toString(),
      customerAttachmentsEnabled:
          json['customer_attachments_enabled'] as bool? ?? false,
      requesterName: json['requester_name']?.toString(),
      requesterEmail: json['requester_email']?.toString(),
      assignedMasterUserName: json['assigned_master_user_name']?.toString(),
      assignedMasterUserEmail: json['assigned_master_user_email']?.toString(),
      lastMessageAt: DateTime.parse(json['last_message_at'].toString()),
      createdAt: DateTime.parse(json['created_at'].toString()),
      updatedAt: DateTime.parse(json['updated_at'].toString()),
      messages: ((json['messages'] as List<dynamic>?) ?? const [])
          .map((item) => MasterSupportMessage.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class MasterSupportTicketInput {
  const MasterSupportTicketInput({
    required this.module,
    required this.priority,
    required this.subject,
    required this.description,
    this.attachmentUrl,
    this.attachmentName,
  });

  final String module;
  final String priority;
  final String subject;
  final String description;
  final String? attachmentUrl;
  final String? attachmentName;

  Map<String, dynamic> toJson() => {
        'module': module,
        'priority': priority,
        'subject': subject,
        'description': description,
        if (attachmentUrl != null) 'attachment_url': attachmentUrl,
        if (attachmentName != null) 'attachment_name': attachmentName,
      };
}
